File: elixir/list_ops.ex
defmodule ListOps do
  @spec count(list) :: non_neg_integer
  def count(l), do: count(l, 0)
  defp count([], ct), do: ct
  defp count([_ | xs], ct), do: count(xs, ct + 1)

  @spec reverse(list) :: list
  def reverse(l), do: reverse(l, [])
  defp reverse([], r), do: r
  defp reverse([x | xs], r), do: reverse(xs, [x | r])

  @spec map(list, (any -> any)) :: list
  def map([], _f), do: []
  def map([x | xs], f), do: [f.(x) | map(xs, f)]

  @spec filter(list, (any -> as_boolean(term))) :: list
  def filter([], _f), do: []

  def filter([x | xs], f) do
    if f.(x), do: [x | filter(xs, f)], else: filter(xs, f)
  end

  @type acc :: any
  @spec foldl(list, acc, (any, acc -> acc)) :: acc
  def foldl([], acc, _f), do: acc

  def foldl([x | xs], acc, f) do
    foldl(xs, f.(x, acc), f)
  end

  @spec foldr(list, acc, (any, acc -> acc)) :: acc
  def foldr([], acc, _f), do: acc

  def foldr([x | xs], acc, f) do
    f.(x, foldr(xs, acc, f))
  end

  @spec append(list, list) :: list
  def append(a, b) do
    foldr(a, b, &[&1 | &2])
  end

  @spec concat([[any]]) :: [any]
  def concat([]), do: []

  def concat(ll) do
    foldr(ll, [], &append/2)
  end
end

File: dart/list_ops.dart
extension ListOps on List {
  void append<T>(List<T> other) {
    for (final each in other) {
      this.add(each);
    }
  }

  List<T> concat<T>() {
    final List<T> arr = [];
    for (final each in this) {
      arr.addAll(each);
    }
    return arr;
  }

  List<T> filter<T>(bool Function(T elem) predicate) {
    final List<T> arr = [];
    for (final each in this) {
      if (predicate(each)) {
        arr.add(each);
      }
    }
    return arr;
  }

  int count() {
    return this.length;
  }

  List<T> myMap<T>(T Function(T elem) fn) {
    final List<T> arr = [];
    for (final each in this) {
      arr.add(fn(each));
    }
    return arr;
  }

  U foldl<T, U>(U initial, U Function(U acc, T elem) fn) {
    for (final each in this) {
      initial = fn(initial, each);
    }
    return initial;
  }

  U foldr<T, U>(U initial, U Function(T elem, U acc) fn) {
    for (final each in this.reversed) {
      initial = fn(each, initial);
    }
    return initial;
  }

  List<T> reverse<T>() {
    final List<T> arr = [];
    for (final each in this.reversed) {
      arr.add(each);
    }
    return arr;
  }
}
